substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use super::{CodeOf, ContractAddressFor, Module, Trait};
+use account_db::{AccountDb, OverlayAccountDb};
+use gas::GasMeter;
+use vm;
+
+use rstd::prelude::*;
+use runtime_primitives::traits::Zero;
+use runtime_support::StorageMap;
+use staking;
+use system;
+
+pub struct CreateReceipt<T: Trait> {
+	pub address: T::AccountId,
+}
+
+pub struct CallReceipt {
+	pub return_data: Vec<u8>,
+}
+
+pub struct ExecutionContext<'a, T: Trait + 'a> {
+	// typically should be dest
+	pub self_account: T::AccountId,
+	pub overlay: OverlayAccountDb<'a, T>,
+	pub depth: usize,
+}
+
+impl<'a, T: Trait> ExecutionContext<'a, T> {
+	/// Make a call to the specified address.
+	pub fn call(
+		&mut self,
+		caller: T::AccountId,
+		dest: T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		_data: Vec<u8>,
+	) -> Result<CallReceipt, &'static str> {
```

Comment:
yeah, I think that would be totally useful! However, I would rather do it in the following PRs.

Reply:
I added it to the list of areas for improvement in the description of PR. 

P.S.: to clarify a bit. We don't actually care that we lose stack trace from `transfer` (since the errors are all different and it's easy to see where it comes from), the thing that bothers me is that there could be very deep nesting of `ExecutionContext`s, and it would be totally cool to see eventually in which `ExecutionContext` the actual error happens.